src/ceph-volume/ceph_volume/devices/raw/list.py

```diff
@@ -27,64 +50,91 @@ def __init__(self, argv):
+        out, err, ret = process.call(lsblk_command)
+        lsblk_json = json.loads(''.join(out))
         result = {}
-        for dev in devs:
+        for dev_info in lsblk_json['blockdevices']:
+            dev = dev_info['name']
             logger.debug('Examining %s' % dev)
-            # bluestore?
-            out, err, ret = process.call([
-                'ceph-bluestore-tool', 'show-label',
-                '--dev', dev], verbose_on_failure=False)
-            if ret:
-                logger.debug('No label on %s' % dev)
-                continue
-            oj = json.loads(''.join(out))
-            if dev not in oj:
-                continue
-            if oj[dev]['description'] != 'main':
-                # ignore non-main devices, for now
+
+            # Because children could be phantom Atari partitions mentioned above, first check if the
+            # parent is a valid bluestore OSD. If it is, the children will only exist if they are
+            # phantom Atari partitions and should be ignored. If it isn't, then the children could
+            # be individual bluestore OSDs.
+            is_bluestore = False
+            try:
+                if disk.has_bluestore_label(dev):
+                    is_bluestore = True
+            except OSError as e:
+                logger.error(('failed to determine if device {} is BlueStore.'.format(dev),
+                             'device children will not be examined to avoid reporting invalid BlueStore data from phantom Atari partitions.',
+                             'err: {}'.format(e)))
                 continue
-            whoami = oj[dev]['whoami']
-            result[oj[dev]['osd_uuid']] = {
-                'type': 'bluestore',
-                'osd_id': int(whoami),
-                'osd_uuid': oj[dev]['osd_uuid'],
-                'ceph_fsid': oj[dev]['ceph_fsid'],
-                'device': dev
-            }
+
+            if is_bluestore:
+                bs_info = _get_bluestore_info(dev)
+                if bs_info is None:
+                    logger.error('failed to get BlueStore info from disk {} with a BlueStore header'.format(dev))
+                    continue
+                result[bs_info['osd_uuid']] = bs_info
+                continue # don't examine children! they could be phantom Atari partitions!
+
+            # When examining children, it isn't necessary to specifically check for the bluestore
+            # label since a failure of the 'ceph-bluestore-tool' command will just mean the child
+            # isn't reported but won't cause us to consider any additional children that might be
+            # invalid Atari partitions.
+            if 'children' in dev_info:
+                for child in dev_info['children']:
+                    child_dev = child['name']
+                    bs_info = _get_bluestore_info(child_dev)
+                    if bs_info is not None:
+                        result[bs_info['osd_uuid']] = bs_info
```

Comment:
We also need to log for the child just like we logged at L122 and `continue` the loop, right?

Reply:
Logging sounds good, but I don't see why we would continue the loop. We need to examine all children for OSDs. Like if, for example, someone had OSDs on partitions 1 and 3 but not 2.

Reply:
Also, logging will happen in the `_get_bluestore_info` function. We can't assume an error here unless we know for sure the disk has a bluestore header, which `_get_bluestore_info` can't really determine. It doesn't change the behavior of this loop to know if it has the 22-byte bluestore header since either way, the disk just gets skipped in reporting.